cpp/src/Ice/Reference.cpp

```diff
@@ -44,91 +42,63 @@ IceInternal::Reference::getCommunicator() const
 ReferencePtr
 IceInternal::Reference::changeContext(const Context& newContext) const
 {
-    ReferencePtr r = _instance->referenceFactory()->copy(this);
+    ReferencePtr r = clone();
     r->_context = new SharedContext(newContext);
     return r;
 }
 
 ReferencePtr
 IceInternal::Reference::changeMode(Mode newMode) const
 {
-    if(newMode == _mode)
```

Comment:
The implementation of the _Xxx function on ObjectPrx already makes this check. No need to do it twice. If you call changeXxx, you get a new reference no matter what.

Unfortunately, this simplification is not sufficient to remove enable_shared_from_this.